src/boot/efi/efi-string.c

```diff
@@ -236,8 +229,45 @@ static bool efi_fnmatch_internal(const char16_t *p, const char16_t *h, int max_d
                 }
 }
 
+/* Patterns are fnmatch-compatible (with reduced feature support). */
 bool efi_fnmatch(const char16_t *pattern, const char16_t *haystack) {
-        return efi_fnmatch_internal(pattern, haystack, 32);
+        /* Patterns can be considered as simple patterns (without '*') concatenated by '*'. By doing so we
+         * simply have to make sure the very first simple pattern matches the start of haystack. Then we just
+         * look for the remaining simple patterns *somewhere* within the haystack (in order) as any extra
+         * characters in between would be matches by the '*'. We then only have to ensure that the very last
+         * simple pattern matches at the actual end of the haystack.
```

Comment:
what does glibc's implementation actually do?

Reply:
I think it does something similar. It would have to do that or the testcase would take forever.

As per usual the code is… difficult to read: https://sourceware.org/git/?p=glibc.git;a=blob;f=posix/fnmatch_loop.c;h=9445ed9c585daafe2ae930d27a6ce17ccff700cd;hb=HEAD

Though, unlike the printf code it is readable (I cannt decide if this is just obfuscation or auto-generated): https://sourceware.org/git/?p=glibc.git;a=blob;f=stdio-common/vfprintf-internal.c;h=fb94961f37265940b19f909901a10d89672988e9;hb=HEAD